src/t_stream.c

```diff
@@ -1338,6 +1348,127 @@ robj *createObjectFromStreamID(streamID *id) {
+int streamIsContiguousRange(stream *s, streamID *start) {
+    streamID start_id, first_id;
+
+    if (!s->length || streamIDEqZero(&s->xdel_max_id)) {
+        /* An empty stream or no XDELs. */
+        return 1;
+    }
+
+    /* Copy start ID, if given, or default to 0-0. */
+    if (start) {
+        start_id = *start;
+    } else {
+        start_id.ms = 0;
+        start_id.seq = 0;
+    }
+
+    streamGetTipID(s,&first_id,1);
+    int cmp_first = streamCompareID(&first_id,&s->xdel_max_id);
+    if (cmp_first == 1) {
+        /* The XDEL is before the first entry. */
+        return 1;
+    }
+
+    int cmp_start = streamCompareID(&start_id,&s->xdel_max_id);
+    if (cmp_start != -1) {
+        /* The range doesn't include an XDEL. */
+        return 1;
+    }
+
+    /* The range includes an XDEL. */
+    return 0;
+}
+
+/* A helper for getting an offset for a given ID in the stream.
+ * A non-zero offset can be either logical (one before the first ID in the
+ * in the stream) or exact (the first or last IDs), but in both cases
+ * it provides a valid lag estimate. The zero offset usually means that
+ * the offset isn't available, except in the case of the 
+ * newly-initialized stream. */
+uint64_t streamGetOffset(stream *s, streamID *id) {
+    /* The offset of any ID in an empty, never-before-used stream is 0. */
+    if (!s->offset) {
+        return 0;
+    }
+
+    /* In the empty stream, if the ID is smaller or equal to the last ID,
+     * it can set to the logical start offset, which is the current offset. */
+    if (!s->length && streamCompareID(id,&s->last_id) < 1) {
+        return s->offset;
+    }
+
+    int cmp_last = streamCompareID(id,&s->last_id);
+    if (cmp_last == 0) {
+        /* Return the exact offset of the last entry in the stream. */
+        return s->offset;
+    } else if (cmp_last == 1) {
+        /* The offset of a future ID is unknown. */
+        return 0;
+    }
+
+    streamID first_id;
+    streamGetTipID(s,&first_id,1);
+    int cmp_first = streamCompareID(id,&first_id);
+    int cmp_xdel = streamCompareID(id,&s->xdel_max_id);
+    if (streamIDEqZero(&s->xdel_max_id) || cmp_xdel == 1) {
+        /* There's definitely no fragmentation. */
+        if (cmp_first == -1) {
+            /* Return the logical start offset. */
+            return s->offset - s->length;
+        } else if (cmp_first == 0) {
+            /* Return the exact offset of the first entry in the stream. */
+            return s->offset - s->length + 1;
+        } else {
+            /* The offset of an arbitrary ID in the stream is unknown. */
+            return 0;
+        }
+    }
+
+    /* There's a chance that the next ID is valid, and the only thing
+     * before the ID is an XDEL. */
+    streamIterator si;
+    int64_t numfields;
+    streamID next_id = *id;
+    streamIncrID(&next_id);
+    streamIteratorStart(&si,s,&next_id,NULL,0);
+    streamIteratorGetID(&si,&next_id,&numfields);   /* Next entry must exist. */
```

Comment:
maybe extend that comment to explain why the next entry must exist (because some of the previous checks validated that we're not looking for something in the far future or the past?), and add an assertion?

Reply:
Comment improved, assertion added.